File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    // bottom-up
    int floor = triangle.size() - 1;
    
    while (floor > 0) {
        for (int i = 0; i < floor; i++) {
            triangle[floor-1][i] += max(triangle[floor][i], triangle[floor][i+1]);
        }
        floor--;
    }    
    
    return triangle[0][0];
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int floor = triangle.length - 1;
        
        while (floor > 0) {
            for (int i = 0; i < floor; i++) {
                triangle[floor-1][i] += Integer.max(triangle[floor][i], triangle[floor][i+1]);
            }
            floor--;
        }
        
        return triangle[0][0];
    }
}